WIP(ish): Real documentation.

fredrikekre opened the pull request:
![screenshot from 2018-09-22 00-21-16](https://user-images.githubusercontent.com/11698744/45908562-790beb80-bdfd-11e8-98c1-5e0db172dd7a.png)

KristofferC commented:
This is great! Thanks a lot for starting this.

timholy commented on docs/src/compatibility.md:
```diff
@@ -0,0 +1,69 @@
+As an example, a version specifier is given as e.g. `1.2.3` is therefore assumed to be compatible with the versions `[1.2.3 - 2.0.0)` where `)` is a non-inclusive upper bound.
+More specifically, a version specifier is either given as a **caret specifier**, e.g. `^1.2.3`  or a **tilde specifier** `~1.2.3`.
+Caret specifiers are the default and hence `1.2.3 == ^1.2.3`. The difference between a caret and tilde is described in the next section.
+The intersection of multiple version specifiers can be formed by comma separating indiviual version specifiers.
```
Do you mean `union` rather than `intersection`? For example if I want to support julia 0.7 and 1.0 I can write 
```julia
 [compat]
julia = "0.7, 1.0"
```
I think this is worth emphasizing with an example, given the semver rules specially about version 0 (making the caret etc less useful) and the fact that so many packages currently use 0.x versions.

Also `JuliaRegistries/General` uses a dash syntax, is that a special case or valid here too?

timholy commented on docs/src/creating-packages.md:
```diff
@@ -0,0 +1,189 @@
+
+### Generating files for a package
+
+To generate files for a new package, use `pkg> generate`.
```
May want to advertise PkgDev in this section, since the "packages" generated this way are kind of limited.

timholy commented on docs/src/creating-packages.md:
````diff
@@ -0,0 +1,189 @@
+ ...
+```
+
+Both `Random` and `JSON` got added to the project’s `Project.toml` file, and the resulting dependencies got added to the `Manifest.toml` file.
````
Somewhere it might be good to have some guidance about whether `Manifest.toml` should be committed to git.

timholy commented on docs/src/creating-packages.md:
````diff
@@ -0,0 +1,189 @@
+A package is built by executing the file `deps/build.jl`.
+
+```
+shell> cat deps/build.log
````
AFAICT `deps` doesn't even exist as a directory at this point, so for anyone following along it will fail. And one should illustrate by creating `build.jl`?

timholy commented on docs/src/managing-packages.md:
```diff
@@ -0,0 +1,298 @@
+Packages can be removed from the current project by using `pkg> rm Package`.
+This will only remove packages that exist in the project, to remove a package that only
+exists as a dependency use `pkg> rm --manifest DepPackage`.
+Note that this will remove all packages that depends on `DepPackage`.
```
Do `resolve` and/or `gc` automatically remove dependencies that are no longer needed?

fredrikekre commented:
Thanks for the comments Tim! I just want to clarify that I have not written any of this, I just split up the existing docs into a more structured (?) representation. We should of course fix the things you pointed out though.

StefanKarpinski commented:
Should we just merge this so that it doesn't keep going stale?

fredrikekre commented:
bors r+

fredrikekre commented:
Hmmm
```
ERROR: Unsatisfiable requirements detected for package Pkg [bfd885c5]:
 Pkg [bfd885c5] log:
 ├─Pkg [bfd885c5] has no known versions!
 └─restricted to versions * by an explicit requirement — no versions left
Stacktrace:
 [1] check_constraints(::Pkg.GraphType.Graph) at /home/travis/build/JuliaLang/Pkg.jl/src/GraphType.jl:935
 [2] Pkg.GraphType.Graph(::Dict{Base.UUID,Set{VersionNumber}}, ::Dict{Base.UUID,Dict{Pkg.Types.VersionRange,Dict{String,Base.UUID}}}, ::Dict{Base.UUID,Dict{Pkg.Types.VersionRange,Dict{String,Pkg.Types.VersionSpec}}}, ::Dict{Base.UUID,String}, ::Dict{Base.UUID,Pkg.Types.VersionSpec}, ::Dict{Base.UUID,Pkg.Types.Fixed}, ::Bool) at /home/travis/build/JuliaLang/Pkg.jl/src/GraphType.jl:362
 [3] deps_graph(::Pkg.Types.Context, ::Dict{Base.UUID,String}, ::Dict{Base.UUID,Pkg.Types.VersionSpec}, ::Dict{Base.UUID,Pkg.Types.Fixed}) at /home/travis/build/JuliaLang/Pkg.jl/src/Operations.jl:307
 [4] macro expansion at ./logging.jl:318 [inlined]
 [5] resolve_versions!(::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}, ::Nothing) at /home/travis/build/JuliaLang/Pkg.jl/src/Operations.jl:353
 [6] resolve_versions! at /home/travis/build/JuliaLang/Pkg.jl/src/Operations.jl:316 [inlined]
 [7] up(::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /home/travis/build/JuliaLang/Pkg.jl/src/Operations.jl:1244
 [8] #up#29(::Pkg.Types.UpgradeLevel, ::Pkg.Types.PackageMode, ::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:198
 [9] up at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:171 [inlined]
 [10] #up#24 at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:163 [inlined]
 [11] up at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:163 [inlined]
 [12] #instantiate#59(::Nothing, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Pkg.Types.Context) at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:531
 [13] #instantiate#58 at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:529 [inlined]
 [14] instantiate() at /home/travis/build/JuliaLang/Pkg.jl/src/API.jl:527
 [15] top-level scope at none:0
```
@KristofferC ?

fredrikekre commented:
bors r+
